### PR TITLE
Take some of the magic out of traits by insisting traits be instances, not classes

### PR DESCRIPTION
**Pull request by jasongrout**
As Sylvain proposed in #48,

```python
class Foo(HasTraits):
    bar = Int    # deprecated
    baz = Int()  # ok
    alpha = List(trait=Int)     # deprecated
    alpha = List(trait=Int())   # ok
```

### CODE REVIEW

**Comment by jasongrout** [comment]
This will require quite a few changes to other repos, because things like `List(Unicode, ...)` are used a lot throughout IPython.

File: traitlets/traitlets.py [hunk 1]
```diff
@@ -64,7 +64,8 @@
 #-----------------------------------------------------------------------------
 # Basic classes
 #-----------------------------------------------------------------------------
-
+import warnings
+warnings.filterwarnings('error')
```
Comment (jasongrout) [inline traitlets/traitlets.py]: This line, changing warnings to errors, should be removed for the final version of this PR.  This just makes it really easy to see where things break.

File: traitlets/traitlets.py [hunk 2]
```diff
@@ -1512,6 +1514,7 @@ def __init__(self, trait=None, default_value=None, **metadata):
             further keys for extensions to the Trait (e.g. config)
 
         """
+
```
Comment (jasongrout) [inline traitlets/traitlets.py]: empty line to be removed

**Comment by SylvainCorlay** [comment]
Jason reviews his own pull requests :)

I am :+1: on this change. This deprecates half of the code of the metaclass.

**Comment by jasongrout** [comment]
Hey, I write them with my right brain, I review them with my left!

**Comment by jasongrout** [comment]
Yes, I think the deprecation should happen in 4.1, removal in 5.0

**Comment by minrk** [comment]
Sounds good.

**Comment by ellisonbg** [comment]
I am +1 on this change.

On Thu, Jul 16, 2015 at 3:11 PM, Min RK <notifications@github.com> wrote:

> Sounds good.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/traitlets/pull/51#issuecomment-122119058>.
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by jasongrout** [comment]
I think this (and the related commits in different repos noted above) are ready for review now.

**Comment by jasongrout** [comment]
ping on merging this.  Because both this and the metadata PR change the same lines when upgrading our code, I'd like to base the metadata pr on this one, and that's easier if this is merged into master.

**Comment by SylvainCorlay** [comment]
cool!

**Pull request merged**